### ISSUE TITLE
Wrappers not working as expected with PISCES 1D parameter sensitiviy analysis workflow, AS 3.14.0

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#941]
In GitLab by @mgali on Feb 6, 2023, 12:36

This is the follow-up of another issue:

[Modified workflow to construct a surrogate model for sensitivity analyses](https://earth.bsc.es/gitlab/cp/genetic_algorithm_pisces1d/-/issues/10)

It may also be related to 

[Horizontal-vertical wrapper fine tuning, autosubmit issue #640](https://earth.bsc.es/gitlab/es/autosubmit/-/issues/640)

I copy the previous message below:

Hi @dbeltrankyl and @mcastril ,

Update:

- I launched **a45b** (account bsc32882) following your specifications, debug queue. 
- I launched **a5mr** (account bsc32340) without vertical wrappers (as I used to do), bsc_es queue.

In both cases I see some chunks were completed, but most failed. Some before completing a single SIM, others after completing a few.

Failures seem related to hitting the wallclock, although I cannot always see the TIME LIMIT message in the logs. I also tried to change the jobs wallclock from 10 to 20 minutes (**a44w** on bsc32882). Same issues. In summary: it seems to me than when jobs in a chunk start running before hitting the wallclock, some of them make it to the target (not all). If the first sim hits the wallclock, then the whole chunk is failed.

I noted some different behaviours compared to previous runs with AS 3.13.0 (Nov 2021): this time I see jobs marked as running but they are actually not! I know because in the runtime folders nothing is happening that indicates a running sim. Just a few jobs are simultaneously queueing, where I would expect all 100 jobs (members) to run or queue simultaneously.

We can assume the nemo simulation environment and executable themselves are fine: I changed nothing compared to older experiments, and I just managed to complete another experiment (**a5mj**) with 2000 sequential sims without failures.

I'll copy this to autosubmit.

**Comment by kinow** [comment]
In GitLab by @mgali on Feb 6, 2023, 12:46

mentioned in issue cp/genetic_algorithm_pisces1d#10

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 6, 2023, 12:46

Hello, 

I think there is something wrong with the template itself. When I look at the logs of each chunk ( from the autosubmit GUI, you can click to each job and see it's log ) I see this:


/esarchive/autosubmit/a5mr/tmp/LOG_a5mr/a5mr_19600101_fc0044_1_SIM.20230203174754.out 

The error is the same for each member I revised


```
ew = pd.DataFrame(data = names.iloc[[row]].values, columns = names.keys(), index = [row])
new.to_csv(str(row)+'_params.csv')
END
++ python - /gpfs/scratch/bsc32/bsc32340/pisces1d/parameters/sobolSeq_N_10000_P_9_2021-11-24_sims2001_3500.csv /gpfs/scratch/bsc32/bsc32340/a5mr/EXP00/0042/namelist_pisces_cfg 0042
Traceback (most recent call last):
  File "<stdin>", line 10, in <module>
  File "/apps/PYTHON/3.6.1/INTEL/lib/python3.6/site-packages/pandas/io/parsers.py", line 676, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/apps/PYTHON/3.6.1/INTEL/lib/python3.6/site-packages/pandas/io/parsers.py", line 448, in _read
    parser = TextFileReader(fp_or_buf, **kwds)
  File "/apps/PYTHON/3.6.1/INTEL/lib/python3.6/site-packages/pandas/io/parsers.py", line 880, in __init__
    self._make_engine(self.engine)
  File "/apps/PYTHON/3.6.1/INTEL/lib/python3.6/site-packages/pandas/io/parsers.py", line 1114, in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)
  File "/apps/PYTHON/3.6.1/INTEL/lib/python3.6/site-packages/pandas/io/parsers.py", line 1891, in __init__
    self._reader = parsers.TextReader(src, **kwds)
  File "pandas/_libs/parsers.pyx", line 374, in pandas._libs.parsers.TextReader.__cinit__
  File "pandas/_libs/parsers.pyx", line 674, in pandas._libs.parsers.TextReader._setup_parser_source
FileNotFoundError: [Errno 2] File /gpfs/scratch/bsc32/bsc32340/pisces1d/parameters/sobolSeq_N_10000_P_9_2021-11-24_sims2001_3500.csv does not exist: '/gpfs/scratch/bsc32/bsc32340/pisces1d/parameters/sobolSeq_N_10000_P_9_2021-11-24_sims2001_3500.csv'
+ nls=
```

Edit: I manually revised it, and indeed this doesn't exists:

```
bsc32070@login2:~> cat /gpfs/scratch/bsc32/bsc32340/pisces1d/parameters/sobolSeq_N_10000_P_9_2021-11-24_sims2001_3500.csv
cat: /gpfs/scratch/bsc32/bsc32340/pisces1d/parameters/sobolSeq_N_10000_P_9_2021-11-24_sims2001_3500.csv: No such file or directory
```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 6, 2023, 12:53

I think there is a typo in the template since:

This doesn't exist 

`cat: /gpfs/scratch/bsc32/bsc32340/pisces1d/parameters/sobolSeq_N_10000_P_9_2021-11-24_sims2001_3500.csv: No such file or directory`

This exists: 

`bsc32070@login2:~> ls /gpfs/scratch/bsc32/bsc32340/pisces1d/parameters/sobolSeq_N_10000_P_9_2021-11-24_sims2001-3500.csv
/gpfs/scratch/bsc32/bsc32340/pisces1d/parameters/sobolSeq_N_10000_P_9_2021-11-24_sims2001-3500.csv
`

( "_" char instead of "-" char)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 6, 2023, 13:50

Hello, 

Sorry, seems that the gui Didn't have the last log, 

The last log had this issue:

```
(Requested nodes are busy)
+ mpirun -np 1 ./nemo
srun: Job 27289028 step creation temporarily disabled, retrying (Requested nodes are busy)
srun: Job 27289028 step creation still disabled, retrying (Requested nodes are busy)
srun: Job 27289028 step creation still disabled, retrying (Requested nodes are busy)
srun: Job 27289028 step creation still disabled, retrying (Requested nodes are busy)
srun: Job 27289028 step creation still disabled, retrying (Requested nodes are busy)
srun: Job 27289028 step creation still disabled, retrying (Requested nodes are busy)
srun: Job 27289028 step creation still disabled, retrying (Requested nodes are busy)
srun: error: Unable to create step for job 27289028: Job/step already completing or completed
```

I think it may be related to the node memory, but I have no experience with this. I think  @mcastril may be able to help you better than I.

Perhaps you can try to limit the horizontal (not sure if the vertical, too ) size using the  {min/max }wrapped_{h/v} to solve this issue, but not sure about the limits.

Also, if you tune the PROCESSORS_PER_NODE=X in platforms.conf you can allocate more or less nodes. Usingm ore nodes can also help

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 6, 2023, 13:59

Okay, forgot about the memory I think this maybe the real issue.


The wrapper is allocating  89 nodes using bsc_es queue but only 50 nodes ( 16 in debug) are allowed in bsc_es

| BSC | 50 (2400) | 48 h |
|-----|-----------|------|

If you limit the horizontal size to 50 it should work. MAX_WRAPPED_H = 50.

What is happening is that some jobs are waiting until the nodes are released ( by other jobs of your experiment) and this causes a mess in the wallclock of some jobs as they are counting already as running but they did not start yet due to the nodes being busy.

**Comment by kinow** [comment]
In GitLab by @mgali on Feb 6, 2023, 14:55

Additional note:

In the previous exps, PROCESSORS_PER_NODE=48. But it seems that less than 48 processors were being allocated, otherwise there wouldn't have been the problem with jobs waiting, right? 

I re-launched the experiments with MAX_WRAPPED_H = 3 to try to ensure we are using as many processors as possible in a given node. Not sure what I did is sensible

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 6, 2023, 15:23

Hello @mgali ,

I've doing some tests, in a dummy copy 

> In the previous exps, PROCESSORS_PER_NODE=48. But it seems that less than 48 processors were being allocated, otherwise there wouldn't have been the problem with jobs waiting, right?

You're right. The issue is that you don't have a TASKS = 1 specified in the jobs.conf. This was causing to autosubmit to request 50 nodes

```
[SIM]
TASKS = 1
```

Autosubmit wrapper thinks SIM requires 48 cores instead of 1 if not specified. 

Then in the autosubmit.conf, you now have only a horizontal defined, you can define a horizontal-vertical:

```
[wrapper]
TYPE = horizontal-vertical
JOBS_IN_WRAPPER = SIM
#MACHINEFILES = STANDARD remove this
MAX_WRAPPED_H = 720 or 2400 ( debug vs bsc_es)
```

then in platforms.conf

if you want to use debug:

```
[marenostrum4]
# Queue type. Options: ps, SGE, LSF, SLURM, PBS, eceaccess
TYPE = slurm
HOST = mn1.bsc.es
PROJECT = bsc32
USER = bsc32### <- your user
SCRATCH_DIR = /gpfs/scratch
ADD_PROJECT_TO_HOST = false
QUEUE = debug
PROCESSORS_PER_NODE = 48
MAX_PROCESSORS = 720
MAX_WALLCLOCK = 02:00
```

if you want to use bsc_es

```
[marenostrum4]
# Queue type. Options: ps, SGE, LSF, SLURM, PBS, eceaccess
TYPE = slurm
HOST = mn1.bsc.es
PROJECT = bsc32
USER = bsc32### <- your user
SCRATCH_DIR = /gpfs/scratch
ADD_PROJECT_TO_HOST = false
QUEUE = bsc_es
PROCESSORS_PER_NODE = 48
MAX_PROCESSORS = 2400
MAX_WALLCLOCK = 48:00
```

If you think there can be memory issues, you can reduce `
PROCESSORS_PER_NODE = 24` 

and decrease the max_wrapped_h value


```
[wrapper]
TYPE = horizontal-vertical
JOBS_IN_WRAPPER = SIM
#MACHINEFILES = STANDARD remove this
MAX_WRAPPED_H = 360 or 1200 ( debug vs bsc_es when processor_per_node = 24 )
```

**Comment by kinow** [comment]
In GitLab by @mgali on Feb 7, 2023, 09:17

Great! Launching a new exp now with all your suggestions. Thanks a lot!Great! Launching a new exp now with all your suggestions. Thanks a lot!

**Comment by kinow** [comment]
In GitLab by @mgali on Feb 8, 2023, 14:53

Well, I re-launched a458 but did a stupid thing, and emptied the tmp directory thinking it just had old stuff. So AS cannot retrieve logs now.

I "cloned" a458 into new exp **a5nd**. This is the one you can safely look at. Fingers crossed

**Comment by kinow** [comment]
In GitLab by @mgali on Feb 8, 2023, 16:00

Yes, sorry about the tmp/LOG_, I explained above that I removed it by "accident". That's why I created a new exp a5nd with the same parameters... let's see. Now have to leave, tomorrow morning will check again. Gràcies!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 8, 2023, 16:01

More SIM jobs are being completed, but their start is delayed. Same error. The node is busy, but eventually, they manage to start. 


Okay, I think I figured out the issue, 

The wrapper is not in --exclusive mode, so more users are using your node. When the first level finishes, the next one does not have all resources available because other HPC users are using it.

There is not currently a way to set --exclusive to a wrapper, only at job level and I think it is not enough. 

The solution is then:

* Use horizontal wrapper, up to 48 jobs ( so it uses one node ).
* Use horizontal-vertical with the same settings but increase the wallclock time of the SIM jobs, so they don't fail before they start ( 20 minutes) and use `RETRIALS = 1` so when they be re-launched when they fail by wallclock

For autosubmit 3.15, I'll add the CUSTOM_DIRECTIVE parameter to wrapper so we can use the --exclusive

**Comment by kinow** [comment]
In GitLab by @mgali on Feb 8, 2023, 16:03

Great, thanks!

BTW, in another experiment (a5mr) I used only horizontal wrappers, and this is the one that compelted the fastest. 

What are then the optimal AS parameters for an experiment with horizontal wrappers only? thanks...

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 8, 2023, 16:06

```
[wrapper]
TYPE = horizontal
JOBS_IN_WRAPPER = SIM
MIN_WRAPPED = 2
MAX_WRAPPED = 48
```
```
[SIM]
TASKS = 1
```
```
[MARENOSTRUM4]
PROCESSORS_PER_NODE = 48
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 8, 2023, 20:15

Sorry for the long silence. As said in the other issue, these are very busy times for us :sweat_smile: . You did great debugging work @dbeltrankyl.

Just to say that in the beginning, I had a hard time getting the picture... because I remember Marcus' workflow had a different than usual logic of dependencies. He was using chunks for the horizontal and members in the vertical and I wasn't sure if that's still the case, but it seems it isn't from what I see in the GUI for `a45b`.

One observation. It seems that your SIMs are only taking 3 minutes each... I think that you could increase the `MAX_WRAPPED_V`, don't you think so?

**Comment by kinow** [comment]
In GitLab by @mgali on Feb 9, 2023, 07:05

Hi all,

Thanks @dbeltrankyl for you latest suggestions and @mcastril for joining the conversation! Highly appreciated given how busy you are :smile: . Needless to say, you will be coauthors of the next paper I prepare...

Just a clarification:

- In the workflow this issue refers too, used for systematic exploration of the parameter space for **sensitivity analysis: members are horizontal and chunks are vertical**.

- However, the templates used here are a modification of the **genetic algorithm (GA)** workflow developed by Marcus Falls. **In the original GA workflow chunks were horizontal and members were vertical** (you remembered right @mcastril !). With Marcus we used horizontal wrappers to run all 100 sims of each generation in parallel, whose results were then collected by the POST job before the next 100 sims started. **I am going to start running GA experiments again during the next few weeks**. Some debugging may be needed but I'm sure it will be very effective after all the work we did here with @dbeltrankyl .

Will continue with the testing today and tell you as soon as something requires your attention. I think we are getting there! thanks again

**Comment by kinow** [comment]
In GitLab by @mgali on Feb 9, 2023, 09:25

Well, let's continue. Early this morning I launched

- a5nd with horizontal wrapper parameters suggested above
- a458 with horizontal-vertical wrapper parameters suggested above

Both failed...

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 9, 2023, 09:48

Hello, 

> One observation. It seems that your SIMs are only taking 3 minutes each... I think that you could increase the `MAX_WRAPPED_V`, don't you think so?

Well, by default, it is up to max_wallclock so there is no need

> * a5nd with horizontal wrapper parameters suggested above
> * a458 with horizontal-vertical wrapper parameters suggested above


Yes, it seems that is the same busy error, some jobs run, and some jobs are being canceled due to wallclock.

> * a458 with horizontal-vertical wrapper parameters suggested above

I see that the jobs were running in exclusive mode in the case of the a458. Not sure what the issue may be. I don't notice any problem with the wrapper itself. The parameters seem correct, and also the list of jobs. 

> * a5nd with horizontal wrapper parameters suggested above

And the a5nd has the same issue but doesn't make sense because it uses a horizontal wrapper, and yesterday, all first 25 members of the horizontal wrapper were completed without issues.


One thing is for sure, we can discard that due to other users using the node. 


@mcastril do you think it could be due to the job.cmd itself? maybe the mpirun -np 1 ./nemo is using more resources than the specified due some hidden flag?

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 10, 2023, 19:38

> Thanks @dbeltrankyl for you latest suggestions and @mcastril for joining the conversation! Highly appreciated given how busy you are :smile: . Needless to say, you will be coauthors of the next paper I prepare...

Thanks @mgali, we appreciate that!

We had a separate discussion about this issue in our team meeting yesterday. The wrapper works fine with a dummy experiment (it runs sleep commands instead of real workloads). So it might be something related to the `mpi` environment as @dbeltrankyl points out.

1) In general, users running horizontal or horizontal-vertical wrappers use `machinefiles`. This is to give a hint to the MPI runtime of where each process will be located. Maybe worth to give it a go? I show the Auto-MONARCH case as an example:

https://earth.bsc.es/gitlab/es/auto-monarch/-/blob/master/templates/05b_sim.sh#L804

That line should be added after the `mpirun` and before the rest of arguments, so that `mpirun` reads the `machinefile` (automatically generated by Autosubmit).

2) Another option is to use a special type of horizontal-wrappers for intra-node allocations that use `srun` instead of `mpirun`:

https://autosubmit.readthedocs.io/en/master/userguide/wrappers/index.html#method

If none of these approaches work, @dbeltrankyl would have to copy your experiment and run it while logging in to one of the compute nodes to figure out what's going on.

We are not sure of why this set up worked for Marcus' experiments and doesn't work now. Environment changes? Could be. We have had issues running `DDT debugger` with `mpirun` lately, which was working perfectly in the past, and operations asked us to move to `srun`.

**Comment by kinow** [comment]
In GitLab by @mgali on Feb 11, 2023, 07:10

Hi,

Thanks for taking care of this and for your inputs. I just checked one of the latest experiments and of course found the error mentioned by @dbeltrankyl above. What I find curious is that when I look at job `a5nh_19600101_fc0002_1_SIM`, for example, I see in the runtime directory that it was running. But it was terminated with that error. See 
`/gpfs/scratch/bsc32/bsc32340/a5nh/LOG_a5nh/a5nh_19600101_fc0002_1_SIM.cmd.err.1`.

PROPOSED PLAN

I am going to run this workflow only a few times. So perhaps it is better that I run it with suboptimal wrappers or even sequentially (worst case scenario, will take just a week) to not abuse of your time.

I will get back to you in 1-2 weeks when I test the **genetic algorithm** workflow. I will have to run several genetic algorithm experiments and in this case it will be more important that the workflow runs smoothly. Sounds good?

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 15, 2023, 17:38

> I am going to run this workflow only a few times. So perhaps it is better that I run it with suboptimal wrappers or even sequentially (worst case scenario, will take just a week) to not abuse of your time.

I'd say that you can try running with:

```
[marenostrum4]
PROCESSORS_PER_NODE = 24
```

As @dbeltrankyl said. 

I don't think it's totally inefficient. NEMO is a memory-bound code and I think that using 24 cores per node, your runs should finish a little bit earlier, as they would have more memory bandwidth per core.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 2, 2023, 10:14

marked this issue as related to #949

**Comment by kinow** [comment]
In GitLab by @mgali on Mar 17, 2023, 11:52

mentioned in issue cp/genetic_algorithm_pisces1d#11
